src/main/java/appeng/util/helpers/P2PHelper.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * This file is part of Applied Energistics 2.
+ * Copyright (c) 2013 - 2015, AlgorithmX2, All rights reserved.
+ *
+ * Applied Energistics 2 is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Applied Energistics 2 is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Applied Energistics 2.  If not, see <http://www.gnu.org/licenses/lgpl>.
+ */
+
+package appeng.util.helpers;
+
+
+import com.google.common.base.Preconditions;
+
+import appeng.api.util.AEColor;
+
+
+public class P2PHelper
+{
+
+	public AEColor[] toColors( short frequency )
+	{
+		final AEColor[] colours = new AEColor[4];
+
+		for( int i = 0; i < 4; i++ )
+		{
+			int nibble = ( frequency >> 4 * ( 3 - i ) ) & 0xF;
+
+			colours[i] = AEColor.values()[nibble];
+		}
+
+		return colours;
+	}
+
+	public short fromColors( AEColor[] colors )
+	{
+		Preconditions.checkArgument( colors.length == 4 );
+
+		int t = 0;
+
+		for( int i = 0; i < 4; i++ )
+		{
+			int code = colors[3 - i].ordinal() << 4 * i;
+
+			t |= code;
+		}
+
+		return (short) ( t & 0xFFFF );
```

Comment:
As `ordinal()` returns values between 0 and 16 (including), `t`can only have bits set outside of 0xFFFF, if `colors[0] == AEColor.TRANSPARENT`. If `colors[1]` to `colors[3]` contain `AEColor.TRANSPARENT`, the higher nibble is potentially messed up, but it will never go beyond the 0xFFFF bitmask. If I understand the idea correctly, we assume that colors will only ever contain one of the 16 "real" colors. In this case, I would argue to remove the bitmask and - if you think it is necessary - check all entries of `colors` against `AEColor.TRANSPARENT`.